osquery/tables/networking/aws_metadata.cpp

```diff
@@ -0,0 +1,238 @@
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. 
+ *
+ */
+
+#include <string>
+
+#include <boost/network/protocol/http/client.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/noncopyable.hpp>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+
+namespace pt = boost::property_tree;
+namespace http = boost::network::http;
+
+namespace osquery {
+namespace tables {
+
+/**
+ * @brief SuperClass for AWS metadata accessors
+ */
+class AwsData : public boost::noncopyable {
+  protected:
+  /**
+   * @brief SQL type for the value
+   */
+  ColumnType sqlType;
+
+  /**
+   * @brief SQL column name
+   */
+  std::string fieldName;
+
+  /**
+   * @brief API URL
+   */
+  std::string subUrl;
+
+  /**
+   * @brief HTTP get the data
+   *
+   * @return osquery Status
+   */
+  std::string DoGet() const;
+
+  public:
+  /**
+   * @brief ctor
+   */
+  AwsData(ColumnType _sqlType, std::string _fieldName, std::string _subUrl)
+    : sqlType(_sqlType), fieldName(std::move(_fieldName)), subUrl(std::move(_subUrl)) {}
+
+  /**
+   * @brief dtor
+   */
+  virtual ~AwsData() {}
+
+  /**
+   * @brief Extract relevant data from return API call, pure virtual
+   *
+   * @param r The row to which the value need to be added
+   * @param http_body content of the http response body
+   * @return osquery Status
+   */
+  virtual Status ExtractResult(Row& r, std::string http_body) const = 0;
+
+  /**
+   * @brief HTTP get and extract data
+   *
+   * @param r The row to which the value need to be added
+   * @return osquery Status
+   */
+  Status Get(Row& r) const {
+      std::string http_body = DoGet();
+      if(!http_body.empty()) {
+        return ExtractResult(r, http_body);
+      }
+      // no hard error if no response
+      return Status(0, "OK");
+  }
+
+};
+
+/**
+ * @brief handle all data not requiring parsing
+ */
+class GenericAwsData : public AwsData {
+  public:
+
+  GenericAwsData(ColumnType in_sqlType, std::string in_fieldName, std::string in_subUrl)
+      : AwsData(in_sqlType, in_fieldName, in_subUrl) {}
+
+  virtual Status ExtractResult(Row& r, std::string http_body) const override;
+};
+
+/**
+ * @brief Handle IAM parsing
+ */
+class IamArnAwsData : public AwsData {
+  public:
+
+  IamArnAwsData(ColumnType in_sqlType, std::string in_fieldName, std::string in_subUrl)
+      : AwsData(in_sqlType, in_fieldName, in_subUrl) {}
+
+  virtual Status ExtractResult(Row& r, std::string http_body) const override;
+};
+
+static bool initialized = false;
+
+// use of the ipv4 address to be region independent (vs hostname based)
+const static std::string kAwsBaseUrl{"http://169.254.169.254/latest/"};
+
+static std::vector<AwsData*> AwsFields; 
+
+void awsMetaInit() {
+    AwsFields.push_back(new GenericAwsData(TEXT_TYPE, "ami_id", "meta-data/ami-id"));
```

Comment:
Some of these are available as JSON with one call:

`curl http://169.254.169.254/latest/dynamic/instance-identity/document`

```
devpayProductCodes
privateIp
availabilityZone
version
pendingTime
instanceId
billingProducts
instanceType
accountId
architecture
kernelId
ramdiskId
imageId
region
```